### PR TITLE
Core/Movement: Implement proper player speed change

### PR DESCRIPTION
**Pull request by chaodhib**
**Changes proposed:**

Implement a part of the changes proposed in https://github.com/TrinityCore/TrinityCore/pull/18771: the speed changes

Taken from that PR, here are the changes visually (only speed changes were made smoother in this PR. Knock backs were kinda implemented already. The other types of changes will come in later PRs):
Before: https://youtu.be/KnsQ_Zqz0kQ
After: https://youtu.be/MwpWViAxhbs

**Target branch(es):** 3.3.5/master

- [X] 3.3.5

**Issues addressed:**

Closes: 
* https://github.com/TrinityCore/TrinityCore/issues/26550
* The issue mentioned [here](https://github.com/TrinityCore/TrinityCore/pull/26538#issuecomment-848529459)
* Partially closes https://github.com/TrinityCore/TrinityCore/issues/18582

**Tests performed:**

Tested IG. No issue found so far.

**Known issues and TODO list:** (add/remove lines as needed)

No known issue. However, the PR is still in WIP. But early feedback and **testing would be highly appreciated** !

EDIT: known issues found:
- [x] trying to reproduce https://github.com/TrinityCore/TrinityCore/issues/21044 results in a kick (with reason: "Took too long to ack a movement change"). EDIT: this is fixed now.

### CODE REVIEW

**Comment by chaodhib** [comment]
@ ckegg @ shenhuyong testing would be highly appreciated!

**Comment by shenhuyong** [comment]
@chaodhib Nice work, I tested it, Judgement of Justice (20184) works fine now.

File: src/server/worldserver/worldserver.conf.dist
Comment (Ovahlord) [inline src/server/worldserver/worldserver.conf.dist:4053]: I think it's more consistent to not use plural here as we don't count anti-cheat as an entity but as a type of mechanic

Comment (chaodhib) [reply src/server/worldserver/worldserver.conf.dist]: Done

**Comment by jackpoz** [comment]
could you please check https://github.com/TrinityCore/TrinityCore/issues/21044 too ? to see if it reappear after this PR.

**Comment by chaodhib** [comment]
> 
> 
> could you please check #21044 too ? to see if it reappear after this PR.

@jackpoz Sure. It's in the "known issue / todo" section of the PR. trying to reproduce results in the player getting kicked. I'm looking into it.

File: src/server/game/Entities/Player/Player.cpp
Comment (Shauren) [inline src/server/game/Entities/Player/Player.cpp:4510]: This seems like a correct change to me (after all this packet handler is responsible for clearing in-flight spell missiles) but this belongs in a separate PR

File: src/server/game/Entities/Unit/Unit.cpp
Comment (Shauren) [inline src/server/game/Entities/Unit/Unit.cpp:56]: Why is this header neccessary here?

**Review by Shauren — CHANGES_REQUESTED** [review]

File: src/server/worldserver/worldserver.conf.dist
Comment (jackpoz) [inline src/server/worldserver/worldserver.conf.dist:4053]: can we have this disabled by default ?

**Review by jackpoz — CHANGES_REQUESTED** [review]

**Comment by Faq** [comment]
could you check also this? https://github.com/TrinityCore/TrinityCore/issues/20136#issuecomment-687663047

File: src/server/game/Entities/Player/Player.cpp
Comment (chaodhib) [reply src/server/game/Entities/Player/Player.cpp]: true. i removed that part

File: src/server/game/Entities/Unit/Unit.cpp
Comment (chaodhib) [reply src/server/game/Entities/Unit/Unit.cpp]: it wasn't. it's the remnants of a rebase I assume. Anyway, I removed it.

**Comment by chaodhib** [comment]
@jackpoz https://github.com/TrinityCore/TrinityCore/issues/21044 remains fixed with this PR. The crash is fixed.

@Faq I checked briefly. The remaining issue doesn't seem to be linked to this PR. Looks like a NPC spawning issue.

File: src/server/worldserver/worldserver.conf.dist
Comment (chaodhib) [reply src/server/worldserver/worldserver.conf.dist:4053]: Maybe we could change this right before the next release. But before that, for the sake of testing, I would suggest to keep this enabled by default. A player getting kicked is an instant feedback that some assumption in the code is wrong or that there is some kind of bug. It would be nice to catch these issues early.

Comment (jackpoz) [reply src/server/worldserver/worldserver.conf.dist:4053]: I'd like to get the issues caused by recent prs fixed first, then we can experiment again with users and ask feedback about new features.

Comment (chaodhib) [reply src/server/worldserver/worldserver.conf.dist:4053]: If i'm not mistaken, all the issues created by https://github.com/TrinityCore/TrinityCore/pull/26348 or https://github.com/TrinityCore/TrinityCore/pull/26538 are either fixed already or they are fixed by this PR. 

However, this PR doesn't just fix issues created recently, it is the first step into making player movement smoother. That's why user feedback on this would be cool.

**Comment by shenhuyong** [comment]
I found a bug, players will be kick when they rides their mounts teleport to a map where mount is not allowed.
`2021-05-29_22:07:58 INFO  [cheat] Unit::CheckPendingMovementAcks: Player GUID: GUID Full: 0x0000000000012509 Type: Player Low: 75017 took too long to acknowledge a movement change. He was therefore kicked.`
How to reproduce it:
- Use command `.go xyz 12890 -7326 66 530`
- Ride a mount
- Go into the instance(Magisters Terrace)
- Then you can see your character moves faster and you will be kick later.

**Comment by chaodhib** [comment]
@shenhuyong thanks! will look into it

File: src/server/worldserver/worldserver.conf.dist
Comment (jackpoz) [reply src/server/worldserver/worldserver.conf.dist:4053]: Would it be ok to leave it disabled by default for some time and then we change it ? we can find other ways to get feedback, like actively asking users to enable the setting, but I would prefer this feature to be opt-in for a while (which would allow to merge this PR earlier as we don't have to test every case).

Maybe like merging the PR, letting 1 month pass with no issues related to it reported, then we can enable it by default ? I can enable it on the test realm too and users can test it there if they want to help.

**Comment by ckegg** [comment]
> @ckegg @shenhuyong testing would be highly appreciated!
Speed Mod issue seems to be resolved after tests.

File: src/server/worldserver/worldserver.conf.dist
Comment (chaodhib) [reply src/server/worldserver/worldserver.conf.dist:4053]: Sure let's do that. I'll make the kick disabled by default, right before merging the PR. I'll keep it enable while the PR is open though, to get as much as feedback as possible, if you don't mind

Comment (jackpoz) [reply src/server/worldserver/worldserver.conf.dist:4053]: is that something that can be implemented with "when AntiCheat.PendingMoveChangesTimeoutTime is set to 0 don't kick, > 0 kick" ?

Comment (chaodhib) [reply src/server/worldserver/worldserver.conf.dist:4053]: Yes, that's what I was thinking too

File: src/server/game/Entities/Unit/Unit.cpp
Comment (ccrs) [inline src/server/game/Entities/Unit/Unit.cpp:12944]: I'm guessing the string can be moved to db

File: src/server/game/Handlers/MovementHandler.cpp
Comment (ccrs) [inline src/server/game/Handlers/MovementHandler.cpp:503]: same in this file

File: src/server/game/Handlers/MovementHandler.cpp
Comment (ccrs) [inline src/server/game/Handlers/MovementHandler.cpp:550]: just a doubt. In case of splines sent by movementgenerator or manually, there will be two UpdatePositions right? one done by the updatespline method and one on movement ack here. right?

File: src/server/game/Movement/MovementPacketSender.cpp
Comment (ccrs) [inline src/server/game/Movement/MovementPacketSender.cpp:87]: maybe this is a nice chance to move it to the new packet format, if you feel like it

File: src/server/game/Entities/Unit/Unit.cpp
Comment (jackpoz) [reply src/server/game/Entities/Unit/Unit.cpp:12944]: that's just a log, we rarely move to db logs

File: src/server/game/Handlers/MovementHandler.cpp
Comment (funjoker) [reply src/server/game/Handlers/MovementHandler.cpp:503]: check jackpoz's response above :)

File: src/server/game/Movement/MovementPacketSender.cpp
Comment (funjoker) [reply src/server/game/Movement/MovementPacketSender.cpp:87]: yeah please do so :)

File: src/server/game/Entities/Unit/Unit.cpp
Comment (illfated) [reply src/server/game/Entities/Unit/Unit.cpp:12944]: The alternative would be to look for an identical string in the OS or external libs, but I guess the chance of finding a close to identical text is pretty slim in generic files.

File: src/server/game/Movement/MovementPacketSender.cpp
Comment (Faq) [reply src/server/game/Movement/MovementPacketSender.cpp:87]: Yeah, but that's another PR.

Comment (ccrs) [reply src/server/game/Movement/MovementPacketSender.cpp:87]: "if you feel like it"

Comment (Ovahlord) [reply src/server/game/Movement/MovementPacketSender.cpp:87]: Nah, since we're using array based packet building here, having a structure build helper here is cleaner that squeezing it all into one class imo. Also out of scope of this PR

File: src/server/game/Entities/Unit/UnitDefines.h
Comment (Ovahlord) [inline src/server/game/Entities/Unit/UnitDefines.h:293]: tabs pls

File: src/server/game/Entities/Unit/UnitDefines.h
Comment (Ovahlord) [inline src/server/game/Entities/Unit/UnitDefines.h:304]: let's move away from defines and go with `static constexpr uint8 MAX_MOVE_TYPE = 9;` instead

File: src/server/game/Entities/Unit/UnitDefines.h
Comment (illfated) [reply src/server/game/Entities/Unit/UnitDefines.h:293]: TC codestyle recommends against Tabs.

Comment (Ovahlord) [reply src/server/game/Entities/Unit/UnitDefines.h:293]: yes and that's no problem when one properly sets up his visual studio to use  tabs4spaces which is what I mean. 
And hi trokli, the stalker of everyone, the one that notices everything and everyone.

Comment (Shauren) [reply src/server/game/Entities/Unit/UnitDefines.h:293]: He didnt mean tabs literally but aligning the = to the same column

Comment (illfated) [reply src/server/game/Entities/Unit/UnitDefines.h:293]: > He didnt mean tabs literally but aligning the = to the same column

Thanks, that interpretation went over my head. I approve.

File: src/server/game/Entities/Unit/UnitDefines.h
Comment (jackpoz) [reply src/server/game/Entities/Unit/UnitDefines.h:304]: the old code had `#define MAX_MOVE_TYPE     9` , let's try to keep as few changes as possible in this PR (and in every PR) so we can focus on the actual issues that it fixes

**Comment by jackpoz** [comment]
it would be nice to update every single line of TC to modern codestyle but let's do that in another PR. This PR is supposed to fix issues that were added in a recent PR, nothing more.

File: src/server/game/Handlers/MovementHandler.cpp
Comment (chaodhib) [reply src/server/game/Handlers/MovementHandler.cpp:550]: Sorry, i'm not sure I follow you. if you could give more context to your question, that would be appreciated

File: src/server/game/Entities/Unit/UnitDefines.h
Comment (chaodhib) [reply src/server/game/Entities/Unit/UnitDefines.h]: done

File: src/server/game/Movement/MovementPacketSender.cpp
Comment (chaodhib) [reply src/server/game/Movement/MovementPacketSender.cpp:87]: so won't do here (but everyone is welcome to improve this in a later PR)

File: src/server/game/Handlers/MovementHandler.cpp
Comment (chaodhib) [reply src/server/game/Handlers/MovementHandler.cpp]: so won't do here (but everyone is welcome to improve this in a later PR)

File: src/server/game/Entities/Unit/Unit.cpp
Comment (chaodhib) [reply src/server/game/Entities/Unit/Unit.cpp:12952]: so, won't do

File: src/server/game/Entities/Unit/UnitDefines.h
Comment (chaodhib) [reply src/server/game/Entities/Unit/UnitDefines.h:304]: so won't do here (but everyone is welcome to improve this in a future PR)

**Comment by chaodhib** [comment]
[This bug](https://github.com/TrinityCore/TrinityCore/pull/26561#issuecomment-850841398) should be fixed now and there is no pending suggestion. If there is nothing else, let's merge this weekend. 

Before I merge, I will change `AntiCheat.PendingMoveChangesTimeoutTime` default value to 0, like said above.

File: src/server/game/Handlers/MovementHandler.cpp
Comment (Ovahlord) [reply src/server/game/Handlers/MovementHandler.cpp:548]: While a unit is being moved by a spline (player charge or knockbacks) the spline system does its own position relocation in the Unit::Update loop. So what he probably asks for is disabling movementInfo based position relocations while the spline is doing its job and use the spline's position instead.

File: src/server/game/Entities/Unit/Unit.cpp
Comment (Faq) [inline src/server/game/Entities/Unit/Unit.cpp:12928]: Why "<"? Timeout can be negative in config?

File: src/server/game/Handlers/MovementHandler.cpp
Comment (ccrs) [reply src/server/game/Handlers/MovementHandler.cpp:548]: not trying to imply any solution myself though, I was just stating the scenario. Don't know what's more appropiate, two updatepositions, the one on updatespline or the one on ack.

Comment (Ovahlord) [reply src/server/game/Handlers/MovementHandler.cpp:548]: IMO we should give the client as little control over such important value as possible so while we do have the chance of using spline data, we should do so let alone to avoid packet spoofing possibilities.

**Comment by jackpoz** [comment]
@chaodhib just as a reminder of how config files work, you'll have to set the value to 0 in worldserver.conf and in the .cpp file where the config is read

**Comment by chaodhib** [comment]
@jackpoz sure, will do. thanks

File: src/server/game/Entities/Unit/Unit.cpp
Comment (chaodhib) [reply src/server/game/Entities/Unit/Unit.cpp:12928]: to be honest, i don't know. will check

File: src/server/game/Handlers/MovementHandler.cpp
Comment (chaodhib) [reply src/server/game/Handlers/MovementHandler.cpp:548]: These 2 lines are in no way different from the lines https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Handlers/MovementHandler.cpp#L396 & https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Handlers/MovementHandler.cpp#L415 in the `HandleMovementOpcodes` handler. The client is allowed to submit positional changes for the units it controls.  The ACK messages carry completely valid positions. The unit did went through these locations (at some point in the past). Disregarding them can actually lead to jerky movements when observing other players.

> the spline system does its own position relocation in the Unit::Update loop

For the majority of the time, spline will only run on units controlled by the server. And here, i'm talking about "real-time client control" (see [this PR](https://github.com/TrinityCore/TrinityCore/pull/26348)). The client is not allowed to send movement messages about them (if it does, they are ignored). So most of the time, there won't be a conflict. Player charges are indeed an exception. But even in retail, charges can lead to jerky movements. It is how the movement system works in WoW. Clients are authoritative on movement for the units under their controls.

PS: Knockback doesn't require a spline (the server can send SMSG_MOVE_KNOCK_BACK to the client, then the physical simulation of the knockback is run on the client).

**Comment by shenhuyong** [comment]
@chaodhib  [This bug](https://github.com/TrinityCore/TrinityCore/pull/26561#issuecomment-850841398) has not been completely fixed.
Another way to reproduce it:
- Use command `.go xyz 3478 266 -120`
- Create a group and convert it to be a raid group.
- Mount a flying mount
- Fly into the instance (Obsidian Sanctum)
- Then you can see your character moves faster and you will be kicked later.

**Comment by chaodhib** [comment]
@shenhuyong Thanks. I'll look into it Monday (I'm not free until then)

**Comment by jackpoz** [comment]
@chaodhib no idea if this is related to the recent changes or not, but if I enter vehicle https://tcubuntu.northeurope.cloudapp.azure.com/aowow/?npc=34775 and exit it without moving further, other players will see me on top of the vehicle:
image

**Comment by chaodhib** [comment]
Sorry for the delay.

@shenhuyong I can't reproduce the bug. For me, it works fine, even with a flying mount, weird. Are you sure you checked out the latest commit of the branch?

@jackpoz I checked and what you are observing is also on the 3.3.5 branch. So while it is clearly a movement bug, it is not related to this PR.

**Comment by Ovahlord** [comment]
Ye, the vehicle exit thing is unrelated to this. It's some bullshit going on in the passenger removal which I think is the spline packet that does this.

**Comment by Aokromes** [comment]
what's missing here to be merged? i got reports of people being slowed after vomit on Festergut

**Comment by chaodhib** [comment]
Nothing as far as I can tell. I'll merge tomorrow evening, if no objection until then.

**Comment by jackpoz** [comment]
could you please apply the "disable the setting by default" with a bit of warning so we can review the final set of changes before merging it 😃 ?

**Comment by chaodhib** [comment]
@jackpoz Done, let me know when I can merge :)

**Comment by jackpoz** [comment]
The default settings look good to me

**Comment by chaodhib** [comment]
Cool, I'm merging. Thank you for the review guys!

File: src/server/game/Handlers/MovementHandler.cpp
Comment (jackpoz) [inline src/server/game/Handlers/MovementHandler.cpp:523]: Please wrap this kick around the config check too so users can disable it if they don't want it

Comment (Aokromes) [reply src/server/game/Handlers/MovementHandler.cpp:523]: 2021-06-16_21:36:17 INFO  [cheat] WorldSession::HandleForceSpeedChangeAck: Player from account id kicked for incorrect data returned in an ack
2021-06-16_21:36:17 INFO  [network.kick] Account: 97003 Character: '' GUID Full: Type: Player Low: kicked with reason: incorrect data returned in an ack

it looks people is being kicked while enter on raid when dead.

**Pull request merged**
